workers/unity/Packages/com.improbable.gdk.mobile/Editor/LaunchMenu.cs

```diff
@@ -70,25 +74,87 @@ private static void LaunchMobileClient()
+        [MenuItem(MenuLaunchiOSDevice, false, 74)]
+        private static void LaunchiOSDeviceClient()
         {
-            foreach (var file in Directory.GetFiles(rootPath, "*.apk", SearchOption.AllDirectories))
+            try
             {
-                apkPath = file;
-                return true;
-            }
+                // Ensure needed tools are installed
+                if (string.IsNullOrEmpty(LibIDeviceInstallerBinary))
+                {
+                    Debug.LogError("Could not find ideviceinstaller tool. Please ensure it is installed. " +
+                        "See https://github.com/libimobiledevice/ideviceinstaller fore more details.");
+                    return;
+                }
+
+                if (string.IsNullOrEmpty(LibIDeviceDebugBinary))
+                {
+                    Debug.LogError("Could not find idevicedebug tool. Please ensure libimobiledevice is installed. " +
+                        "See https://helpmanual.io/help/idevicedebug/ for more details.");
+                    return;
+                }
+
+                EditorUtility.DisplayProgressBar("Launching iOS Device Client", "Installing archive", 0.3f);
+
+                // Get chosen ios package id
+                var bundleId = PlayerSettings.GetApplicationIdentifier(BuildTargetGroup.iOS);
+                string appList = null;
+
+                // Find archive to install
+                var ipaPath = Directory.GetFiles(AbsoluteAppBuildPath, "*.ipa", SearchOption.AllDirectories).FirstOrDefault();
+                RedirectedProcess.Command(LibIDeviceInstallerBinary).WithArgs("-l")
+                    .AddAccumulatedOutputProcessing(output => { appList = output; })
+                    .RedirectOutputOptions(OutputRedirectBehaviour.None).Run();
+                var existsOnDevice = appList.Contains(bundleId);
+                var existsLocally = !string.IsNullOrEmpty(ipaPath);
+                if (!existsLocally && !existsOnDevice)
+                {
+                    Debug.LogError($"Could not find an app on device or built out iOS .ipa archive in \"{AbsoluteAppBuildPath}\" to launch.");
+                }
+
+                if (existsLocally)
+                {
+                    if (RedirectedProcess.Command(LibIDeviceInstallerBinary)
+                            .WithArgs((existsOnDevice ? "-g" : "-i"), $"\"{ipaPath}\"").Run() !=
+                        0)
+                    {
+                        Debug.LogError(
+                            "Error while installing .ipa archive to the device. Please check the log for details about the error.");
+                        return;
+                    }
+                }
 
-            apkPath = string.Empty;
-            return false;
+                // Wait until the app is installed
+                while (RedirectedProcess.Command(LibIDeviceInstallerBinary).WithArgs("-l").Run() != 0)
```

Comment:
this looks sketchy. why do we need to wrap it in a while loop?
is it possible for the user to do something funky, e.g. disconnect the device, close the simulator so that Unity becomes completely unresponsive until you force quit it?

Reply:
Fair point. Let me add a timeout here so that we don't hold off indefinitely

Reply:
Okay, so after some digging in the code I found out that `ideviceinstaller -l`(list) command has waiting for device built in, so we don't have to mess with timeout. Instead we can just call list command and when it returns we're ready to run.
I tested this extensively on slow device with FPS starter project to ensure this works as expected and didn't see any issues.